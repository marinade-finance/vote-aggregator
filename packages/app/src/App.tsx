import {useCallback, useMemo} from 'react';
import {QueryClient, QueryClientProvider} from '@tanstack/react-query';
import {CssBaseline} from '@mui/material';
import {
  Adapter,
  WalletAdapterNetwork,
  WalletError,
} from '@solana/wallet-adapter-base';
import {clusterApiUrl} from '@solana/web3.js';
import {useSnackbar} from 'notistack';
import {ConnectionProvider, WalletProvider} from '@solana/wallet-adapter-react';
import {WalletDialogProvider} from '@solana/wallet-adapter-material-ui';
import Header from './components/Header';
import ClanList from './components/ClanList';

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      staleTime: 1000 * 60,
    },
  },
});

function App() {
  // The network can be set to 'devnet', 'testnet', or 'mainnet-beta'.
  const network = WalletAdapterNetwork.Devnet;

  // You can also provide a custom RPC endpoint.
  const endpoint = useMemo(() => clusterApiUrl(network), [network]);

  const {enqueueSnackbar} = useSnackbar();
  const onError = useCallback(
    (error: WalletError, adapter?: Adapter) => {
      enqueueSnackbar(
        error.message ? `${error.name}: ${error.message}` : error.name,
        {variant: 'error'}
      );
      console.error(error, adapter);
    },
    [enqueueSnackbar]
  );

  return (
    <QueryClientProvider client={queryClient}>
      <CssBaseline />
      <ConnectionProvider endpoint={endpoint}>
        <WalletProvider wallets={[]} onError={onError} autoConnect>
          <WalletDialogProvider>
            <Header />
            <ClanList sx={{width: '100%', ml: 10}} />
          </WalletDialogProvider>
        </WalletProvider>
      </ConnectionProvider>
    </QueryClientProvider>
  );
}

export default App;
