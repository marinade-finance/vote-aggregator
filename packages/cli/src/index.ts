import {cli} from './cli';

cli().parseAsync(process.argv);
