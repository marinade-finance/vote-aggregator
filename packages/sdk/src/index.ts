export * from './sdk';
export * from './root';
export * from './member';
export * from './clan';
