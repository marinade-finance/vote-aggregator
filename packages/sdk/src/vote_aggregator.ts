../../../target/types/vote_aggregator.ts