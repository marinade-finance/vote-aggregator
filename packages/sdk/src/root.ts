import {
  GetProgramAccountsFilter,
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from '@solana/web3.js';
import {RealmSide, VoteAggregatorSdk} from './sdk';
import {
  GoverningTokenConfig,
  MintMaxVoteWeightSource,
  createSetRealmConfig,
  getRealm,
  getRealmConfig,
  getRealmConfigAddress,
} from '@solana/spl-governance';
import {
  AnchorProvider,
  IdlAccounts,
  ProgramAccount,
} from '@coral-xyz/anchor';
import {VoteAggregator} from './vote_aggregator';
import {splGovernanceProgram} from '@coral-xyz/spl-governance';
import BN from 'bn.js';

export type RootAccount = IdlAccounts<VoteAggregator>['root'];
export type MaxVoterWeightAccount =
  IdlAccounts<VoteAggregator>['maxVoterWeightRecord'];

export class RootSdk {
  constructor(public readonly sdk: VoteAggregatorSdk) {}

  rootAddress({
    realmAddress,
    governingTokenMint,
  }: {
    realmAddress: PublicKey;
    governingTokenMint: PublicKey;
  }): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [
        Buffer.from('root', 'utf-8'),
        realmAddress.toBuffer(),
        governingTokenMint.toBuffer(),
      ],
      this.sdk.programId
    );
  }

  maxVoterWieghtAddress({
    rootAddress,
  }: {
    rootAddress: PublicKey;
  }): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [Buffer.from('max-voter-weight', 'utf-8'), rootAddress.toBuffer()],
      this.sdk.programId
    );
  }

  lockAuthority({rootAddress}: {rootAddress: PublicKey}): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [Buffer.from('lock-authority', 'utf-8'), rootAddress.toBuffer()],
      this.sdk.programId
    );
  }

  fetchRoot(rootAddress: PublicKey): Promise<RootAccount> {
    return this.sdk.program.account.root.fetch(rootAddress);
  }

  fetchRoots({
    governanceProgram,
    realm,
    governingTokenMint,
    votingWeightPlugin,
  }: {
    governanceProgram?: PublicKey;
    realm?: PublicKey;
    governingTokenMint?: PublicKey;
    votingWeightPlugin?: PublicKey;
  }): Promise<ProgramAccount<RootAccount>[]> {
    let filter: GetProgramAccountsFilter[] | undefined;

    if (governanceProgram) {
      if (!filter) {
        filter = [];
      }
      filter.push({
        memcmp: {
          offset: 8,
          bytes: governanceProgram.toBase58(),
        },
      });
    }

    if (realm) {
      if (!filter) {
        filter = [];
      }
      filter.push({
        memcmp: {
          offset: 40,
          bytes: realm.toBase58(),
        },
      });
    }

    if (governingTokenMint) {
      if (!filter) {
        filter = [];
      }
      filter.push({
        memcmp: {
          offset: 72,
          bytes: governingTokenMint.toBase58(),
        },
      });
    }

    if (votingWeightPlugin) {
      if (!filter) {
        filter = [];
      }
      filter.push({
        memcmp: {
          offset: 104,
          bytes: votingWeightPlugin.toBase58(),
        },
      });
    }

    return this.sdk.program.account.root.all(filter);
  }

  fetchMaxVoterWeight({
    rootAddress,
    maxVoterWeightAddress,
  }: {
    rootAddress?: PublicKey;
    maxVoterWeightAddress?: PublicKey;
  }): Promise<MaxVoterWeightAccount> {
    if (!maxVoterWeightAddress) {
      if (!rootAddress) {
        throw new Error('rootAddress is required');
      }
      maxVoterWeightAddress = this.maxVoterWieghtAddress({rootAddress})[0];
    }
    return this.sdk.program.account.maxVoterWeightRecord.fetch(
      maxVoterWeightAddress
    );
  }

  async createRootInstructions({
    splGovernanceId,
    realmAddress,
    realmData,
    realmConfigData,
    side,
    maxProposalLifetime = new BN(0),
    payer,
  }: {
    splGovernanceId?: PublicKey;
    realmAddress: PublicKey;
    realmData?: {
      communityMint: PublicKey;
      config: {
        councilMint: PublicKey | undefined;
        communityMintMaxVoteWeightSource: MintMaxVoteWeightSource;
        minCommunityTokensToCreateGovernance: BN;
      };
      authority: PublicKey | undefined;
    };
    realmConfigData?: {
      communityTokenConfig: GoverningTokenConfig;
      councilTokenConfig: GoverningTokenConfig;
    };
    side: RealmSide;
    maxProposalLifetime?: BN;
    payer: PublicKey;
  }): Promise<TransactionInstruction[]> {
    if (!realmData) {
      const {account, owner} = await getRealm(
        this.sdk.connection,
        realmAddress
      );
      if (splGovernanceId && !owner.equals(splGovernanceId)) {
        throw new Error(
          `Realm ${realmAddress} is not owned by ${splGovernanceId}`
        );
      }
      realmData = account;
      splGovernanceId = owner;
    }
    if (!splGovernanceId) {
      throw new Error('splGovernanceId is required');
    }
    if (!realmData.authority) {
      throw new Error(`Realm ${realmAddress} does not have an authority`);
    }

    if (!realmConfigData) {
      const {account} = await getRealmConfig(
        this.sdk.connection,
        await getRealmConfigAddress(splGovernanceId, realmAddress)
      );
      realmConfigData = account;
    }

    const governingTokenMint =
      side === 'community'
        ? realmData.communityMint
        : realmData.config.councilMint!;
    if (!governingTokenMint) {
      throw new Error(`Realm ${realmAddress} does not have a ${side} mint`);
    }

    const [rootAddress] = this.rootAddress({realmAddress, governingTokenMint});
    const [maxVwr] = this.maxVoterWieghtAddress({rootAddress});

    const createRootIx = await this.sdk.program.methods
      .createRoot(maxProposalLifetime)
      .accountsStrict({
        root: rootAddress,
        realm: realmAddress,
        realmConfig: await getRealmConfigAddress(splGovernanceId, realmAddress),
        governingTokenMint,
        realmAuthority: realmData.authority,
        maxVwr,
        payer,
        governanceProgram: splGovernanceId,
        systemProgram: SystemProgram.programId,
        voteAggregatorProgram: this.sdk.program.programId,
      })
      .instruction();

    const communityTokenConfigArgs = {
      voterWeightAddin:
        side === 'community'
          ? this.sdk.program.programId
          : realmConfigData.communityTokenConfig.voterWeightAddin,
      maxVoterWeightAddin:
        realmConfigData.communityTokenConfig.maxVoterWeightAddin,
      tokenType: realmConfigData.communityTokenConfig.tokenType,
      useVoterWeightAddin:
        side === 'community' ||
        Boolean(realmConfigData.communityTokenConfig.voterWeightAddin),
      useMaxVoterWeightAddin: Boolean(
        realmConfigData.communityTokenConfig.maxVoterWeightAddin
      ),
    };
    const councilTokenConfigArgs = realmData.config.councilMint
      ? {
          voterWeightAddin:
            side === 'council'
              ? this.sdk.program.programId
              : realmConfigData.councilTokenConfig.voterWeightAddin,
          maxVoterWeightAddin:
            realmConfigData.councilTokenConfig.maxVoterWeightAddin,
          tokenType: realmConfigData.councilTokenConfig.tokenType,
          useVoterWeightAddin:
            side === 'council' ||
            Boolean(realmConfigData.councilTokenConfig.voterWeightAddin),
          useMaxVoterWeightAddin: Boolean(
            realmConfigData.councilTokenConfig.maxVoterWeightAddin
          ),
        }
      : undefined;

    const setRealmConfigIx = await createSetRealmConfig(
      splGovernanceId,
      3,
      realmAddress,
      realmData.authority!,
      realmData.config.councilMint || undefined,
      realmData.config.communityMintMaxVoteWeightSource,
      realmData.config.minCommunityTokensToCreateGovernance,
      communityTokenConfigArgs,
      councilTokenConfigArgs,
      payer
    );

    // TODO: use official JS SDK
    const splGovernance = splGovernanceProgram({
      programId: splGovernanceId,
      provider: new AnchorProvider(
        this.sdk.connection,
        {
          signTransaction: t => Promise.resolve(t),
          signAllTransactions: ts => Promise.resolve(ts),
          publicKey: PublicKey.default,
        },
        {}
      ),
    });
    const [lockAuthority] = this.lockAuthority({rootAddress});
    const setLockAuthorityIx = await splGovernance.methods
      .setRealmConfigItem({
        tokenOwnerRecordLockAuthority: {
          action: {
            add: {},
          },
          governingTokenMint,
          authority: lockAuthority,
        },
      })
      .accountsStrict({
        realm: realmAddress,
        realmAuthority: realmData.authority!,
        payer,
        systemProgram: SystemProgram.programId,
        realmConfigAddress: await getRealmConfigAddress(
          splGovernanceId,
          realmAddress
        ),
      })
      .instruction();

    return [createRootIx, setRealmConfigIx, setLockAuthorityIx];
  }

  async setMaxProposalLifetimeInstruction({
    maxProposalLifetime,
    root,
    realm,
    realmAuthority,
  }: {
    maxProposalLifetime: BN;
    root: PublicKey;
    realm: PublicKey;
    realmAuthority: PublicKey;
  }) {
    return await this.sdk.program.methods
      .setMaxProposalLifetime(maxProposalLifetime)
      .accountsStrict({
        realm,
        realmAuthority,
        root,
      })
      .instruction();
  }

  async setVoterWeightResetInstruction({
    step,
    nextResetTime,
    root,
    realm,
    realmAuthority,
  }: {
    step: BN;
    nextResetTime: BN | null;
    root: PublicKey;
    realm: PublicKey;
    realmAuthority: PublicKey;
  }) {
    return await this.sdk.program.methods
      .setVoterWeightReset(step, nextResetTime)
      .accountsStrict({
        realm,
        realmAuthority,
        root,
      })
      .instruction();
  }
}
