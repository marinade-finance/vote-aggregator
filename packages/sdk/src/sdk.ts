import {Program, Provider} from '@coral-xyz/anchor';
import {VoteAggregator, IDL} from './vote_aggregator';
import {
  AccountMeta,
  Connection,
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from '@solana/web3.js';
import {
  Realm,
  RealmConfigAccount,
  getRealm,
  getRealmConfig,
  getRealmConfigAddress,
  getTokenHoldingAddress,
} from '@solana/spl-governance';

class ReadonlyProvider implements Provider {
  constructor(public connection: Connection) {}
}

export class VoteAggregatorSdk {
  program: Program<VoteAggregator>;

  constructor(
    connection: Connection = new Connection('http://localhost:8899'),
    programId: PublicKey = new PublicKey(
      'DDN7fpM4tY3ZHdgSuf8B4UBMakh4kqPzuDZHxgVyyNg'
    )
  ) {
    this.program = new Program(
      IDL,
      programId,
      new ReadonlyProvider(connection)
    );
  }

  get programId(): PublicKey {
    return this.program.programId;
  }

  get connection(): Connection {
    return this.program.provider.connection;
  }

  async createRootInstruction({
    splGovernanceId,
    realmId,
    realmData,
    realmConfigData,
    side,
    payer,
  }: {
    splGovernanceId?: PublicKey;
    realmId: PublicKey;
    realmData?: Realm;
    realmConfigData?: RealmConfigAccount;
    side: 'community' | 'council';
    payer: PublicKey;
  }): Promise<TransactionInstruction> {
    if (!realmData) {
      const {account, owner} = await getRealm(this.connection, realmId);
      if (splGovernanceId && !owner.equals(splGovernanceId)) {
        throw new Error(`Realm ${realmId} is not owned by ${splGovernanceId}`);
      }
      realmData = account;
      splGovernanceId = owner;
    }
    if (!splGovernanceId) {
      throw new Error('splGovernanceId is required');
    }
    if (!realmConfigData) {
      const {account} = await getRealmConfig(
        this.connection,
        await getRealmConfigAddress(splGovernanceId, realmId)
      );
      realmConfigData = account;
    }

    const governingTokenMint =
      side === 'community'
        ? realmData.communityMint
        : realmData.config.councilMint!;
    if (!governingTokenMint) {
      throw new Error(`Realm ${realmId} does not have a ${side} mint`);
    }

    const [rootAddress] = PublicKey.findProgramAddressSync(
      [
        Buffer.from('root', 'utf-8'),
        realmId.toBuffer(),
        governingTokenMint.toBuffer(),
      ],
      this.program.programId
    );
    const [maxVoterWeightAddress] = PublicKey.findProgramAddressSync(
      [Buffer.from('max-voter-weight', 'utf-8'), rootAddress.toBuffer()],
      this.program.programId
    );

    const extraAccounts: AccountMeta[] = [
      {
        pubkey: realmData.communityMint,
        isWritable: true,
        isSigner: false,
      },
      {
        pubkey: await getTokenHoldingAddress(
          splGovernanceId,
          realmId,
          realmData.communityMint
        ),
        isWritable: true,
        isSigner: false,
      },
    ];
    if (realmData.config.councilMint) {
      extraAccounts.push({
        pubkey: realmData.config.councilMint!,
        isWritable: true,
        isSigner: false,
      });
      extraAccounts.push({
        pubkey: await getTokenHoldingAddress(
          splGovernanceId,
          realmId,
          realmData.config.councilMint
        ),
        isWritable: true,
        isSigner: false,
      });
    }
    if (realmConfigData.communityTokenConfig.voterWeightAddin) {
      extraAccounts.push({
        pubkey: realmConfigData.communityTokenConfig.voterWeightAddin,
        isWritable: false,
        isSigner: false,
      });
    }
    if (realmConfigData.communityTokenConfig.maxVoterWeightAddin) {
      extraAccounts.push({
        pubkey: realmConfigData.communityTokenConfig.maxVoterWeightAddin,
        isWritable: false,
        isSigner: false,
      });
    }
    if (realmConfigData.councilTokenConfig.voterWeightAddin) {
      extraAccounts.push({
        pubkey: realmConfigData.councilTokenConfig.voterWeightAddin,
        isWritable: false,
        isSigner: false,
      });
    }
    if (realmConfigData.councilTokenConfig.maxVoterWeightAddin) {
      extraAccounts.push({
        pubkey: realmConfigData.councilTokenConfig.maxVoterWeightAddin,
        isWritable: false,
        isSigner: false,
      });
    }

    return await this.program.methods
      .createRoot()
      .accountsStrict({
        root: rootAddress,
        realm: realmId,
        realmConfig: await getRealmConfigAddress(splGovernanceId, realmId),
        governingTokenMint,
        realmAuthority: realmData.authority!,
        maxVoterWeight: maxVoterWeightAddress,
        payer,
        governanceProgram: splGovernanceId,
        systemProgram: SystemProgram.programId,
        voteAggregatorProgram: this.program.programId,
      })
      .remainingAccounts(extraAccounts)
      .instruction();
  }
}
