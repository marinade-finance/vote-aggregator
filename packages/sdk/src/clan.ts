import {PublicKey, SystemProgram} from '@solana/web3.js';
import {VoteAggregatorSdk} from './sdk';
import {IdlAccounts} from '@coral-xyz/anchor';
import {VoteAggregator} from './vote_aggregator';
import {RootAccount} from './root';

export type ClanAccount = IdlAccounts<VoteAggregator>['clan'];
export type VoterWeightAccount =
  IdlAccounts<VoteAggregator>['voterWeightRecord'];

export class ClanSdk {
  constructor(public readonly sdk: VoteAggregatorSdk) {}

  voterAuthority({clanAddress}: {clanAddress: PublicKey}): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [Buffer.from('voter-authority', 'utf-8'), clanAddress.toBuffer()],
      this.sdk.programId
    );
  }

  tokenOwnerRecordAddress({
    realmAddress,
    governingTokenMint,
    clanAddress,
    voterAuthority,
    splGovernanceId,
  }: {
    realmAddress: PublicKey;
    governingTokenMint: PublicKey;
    clanAddress?: PublicKey;
    voterAuthority?: PublicKey;
    splGovernanceId: PublicKey;
  }): [PublicKey, number] {
    if (!voterAuthority) {
      if (!clanAddress) {
        throw new Error('clanAddress is required');
      }
      voterAuthority = this.voterAuthority({clanAddress})[0];
    }
    return PublicKey.findProgramAddressSync(
      [
        Buffer.from('governance', 'utf-8'),
        realmAddress.toBuffer(),
        governingTokenMint.toBuffer(),
        voterAuthority.toBuffer(),
      ],
      splGovernanceId
    );
  }

  voterWeightAddress(clanAddress: PublicKey): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [Buffer.from('voter-weight', 'utf-8'), clanAddress.toBuffer()],
      this.sdk.programId
    );
  }

  fetchClan(clanAddress: PublicKey): Promise<ClanAccount> {
    return this.sdk.program.account.clan.fetch(clanAddress);
  }

  fetchVoterWeight({
    clanAddress,
    voterWeightAddress,
  }: {
    clanAddress?: PublicKey;
    voterWeightAddress?: PublicKey;
  }): Promise<VoterWeightAccount> {
    if (!voterWeightAddress) {
      if (!clanAddress) {
        throw new Error('clanAddress is required');
      }
      voterWeightAddress = this.voterWeightAddress(clanAddress)[0];
    }
    return this.sdk.program.account.voterWeightRecord.fetch(voterWeightAddress);
  }

  async createClanInstruction({
    rootAddress,
    root,
    clanAddress,
    owner,
    payer,
  }: {
    rootAddress: PublicKey;
    root: RootAccount;
    clanAddress: PublicKey;
    owner: PublicKey;
    payer: PublicKey;
  }) {
    const [voterAuthority] = this.voterAuthority({clanAddress});
    const [tokenOwnerRecord] = this.tokenOwnerRecordAddress({
      realmAddress: root.realm,
      governingTokenMint: root.governingTokenMint,
      clanAddress,
      splGovernanceId: root.governanceProgram,
    });
    const [voterWeightRecord] = this.voterWeightAddress(clanAddress);
    return await this.sdk.program.methods
      .createClan(owner)
      .accountsStrict({
        root: rootAddress,
        clan: clanAddress,
        realm: root.realm,
        governingTokenMint: root.governingTokenMint,
        payer,
        governanceProgram: root.governanceProgram,
        systemProgram: SystemProgram.programId,
        voterAuthority,
        tokenOwnerRecord,
        voterWeightRecord,
      })
      .instruction();
  }
}
