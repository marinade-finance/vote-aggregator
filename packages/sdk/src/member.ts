import {PublicKey, SystemProgram} from '@solana/web3.js';
import {VoteAggregatorSdk} from './sdk';
import {BN, IdlAccounts} from '@coral-xyz/anchor';
import {VoteAggregator} from './vote_aggregator';
import {RootAccount} from './root';
import {SYSTEM_PROGRAM_ID} from '@solana/spl-governance';
import {VoterWeightAccount} from './clan';

export type MemberAccount = IdlAccounts<VoteAggregator>['member'];

export class MemberSdk {
  constructor(public readonly sdk: VoteAggregatorSdk) {}

  memberAddress({
    rootAddress,
    owner,
  }: {
    rootAddress: PublicKey;
    owner: PublicKey;
  }): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [
        Buffer.from('member', 'utf-8'),
        rootAddress.toBuffer(),
        owner.toBuffer(),
      ],
      this.sdk.programId
    );
  }

  tokenOwnerRecordAddress({
    realmAddress,
    governingTokenMint,
    owner,
    splGovernanceId,
  }: {
    realmAddress: PublicKey;
    governingTokenMint: PublicKey;
    owner: PublicKey;
    splGovernanceId: PublicKey;
  }): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [
        Buffer.from('governance', 'utf-8'),
        realmAddress.toBuffer(),
        governingTokenMint.toBuffer(),
        owner.toBuffer(),
      ],
      splGovernanceId
    );
  }

  fetchMember({
    memberAddress,
    owner,
    rootAddress,
  }: {
    memberAddress?: PublicKey;
    rootAddress?: PublicKey;
    owner?: PublicKey;
  }): Promise<MemberAccount> {
    if (!memberAddress) {
      if (!owner) {
        throw new Error('owner is required');
      }
      if (!rootAddress) {
        throw new Error('rootAddress is required');
      }
      memberAddress = this.memberAddress({rootAddress, owner})[0];
    }
    return this.sdk.program.account.member.fetch(memberAddress);
  }

  async findVoterWeightRecord({
    root,
    member,
  }: {
    root: RootAccount;
    member: MemberAccount;
  }) {
    const voterWeightAccounts = await this.sdk.connection.getProgramAccounts(
      root.votingWeightPlugin,
      {
        filters: [
          {
            memcmp: {
              offset: 0,
              bytes: '8riZd8mYDQk', // Discriminator
            },
          },
          {
            memcmp: {
              offset: 8,
              bytes: root.realm.toBase58(),
            },
          },
          {
            memcmp: {
              offset: 8 + 32,
              bytes: root.governingTokenMint.toBase58(),
            },
          },
          {
            memcmp: {
              offset: 8 + 32 + 32,
              bytes: member.owner.toBase58(),
            },
          },
        ],
      }
    );

    const voterWeightRecords = voterWeightAccounts.map(({account, pubkey}) => {
      const record = this.sdk.program.coder.accounts.decode<VoterWeightAccount>(
        'voterWeightRecord',
        account.data
      );
      return {
        record,
        pubkey,
      };
    });

    let maxExpiry: BN | null = null;
    let maxPower = new BN(0);
    let bestIndex = null;
    for (let i = 0; i < voterWeightRecords.length; i++) {
      const {record} = voterWeightRecords[i];
      if (record.weightAction || record.weightActionTarget) {
        continue;
      }
      if (
        (!record.voterWeightExpiry && maxExpiry) ||
        (maxExpiry && record.voterWeightExpiry?.gt(maxExpiry))
      ) {
        maxExpiry = record.voterWeightExpiry;
        maxPower = record.voterWeight;
        bestIndex = i;
      } else if (
        (record.voterWeightExpiry === null && maxExpiry === null) ||
        (maxExpiry && record.voterWeightExpiry?.eq(maxExpiry))
      ) {
        if (record.voterWeight.gt(maxPower)) {
          maxPower = record.voterWeight;
          bestIndex = i;
        }
      }
    }

    if (bestIndex === null) {
      throw new Error(
        `Can not find VWR for ${member.owner.toBase58()} in plugin ${root.votingWeightPlugin.toBase58()}`
      );
    }

    return voterWeightRecords[bestIndex];
  }

  async createMemberInstruction({
    rootAddress,
    root,
    owner,
    payer,
  }: {
    rootAddress: PublicKey;
    root: RootAccount;
    owner: PublicKey;
    payer: PublicKey;
  }) {
    const [tokenOwnerRecord] = this.tokenOwnerRecordAddress({
      realmAddress: root.realm,
      governingTokenMint: root.governingTokenMint,
      owner,
      splGovernanceId: root.governanceProgram,
    });
    const [member] = this.memberAddress({rootAddress, owner});
    return await this.sdk.program.methods
      .createMember()
      .accountsStrict({
        root: rootAddress,
        member,
        payer,
        systemProgram: SYSTEM_PROGRAM_ID,
        tokenOwnerRecord,
        owner,
      })
      .instruction();
  }

  async joinClanInstruction({
    root,
    member,
    clanAddress,
    memberVoterWeightAddress,
  }: {
    root: RootAccount;
    member: MemberAccount;
    clanAddress: PublicKey;
    memberVoterWeightAddress: PublicKey;
  }) {
    const [memberAddress] = this.memberAddress({
      rootAddress: member.root,
      owner: member.owner,
    });
    return await this.sdk.program.methods
      .joinClan()
      .accountsStrict({
        root: member.root,
        member: memberAddress,
        clan: clanAddress,
        memberAuthority: member.owner, // TODO delegate
        clanVoterWeightRecord: this.sdk.clan.voterWeightAddress(clanAddress)[0],
        memberTokenOwnerRecord: this.tokenOwnerRecordAddress({
          realmAddress: root.realm,
          governingTokenMint: root.governingTokenMint,
          owner: member.owner,
          splGovernanceId: root.governanceProgram,
        })[0],
        memberVoterWeightRecord: memberVoterWeightAddress,
        maxVoterWeightRecord: this.sdk.root.maxVoterWieghtAddress({
          rootAddress: member.root,
        })[0],
      })
      .instruction();
  }
}
