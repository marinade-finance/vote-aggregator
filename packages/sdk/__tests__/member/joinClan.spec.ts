import {
  ClanTester,
  JoinClanTestData,
  MemberTester,
  RealmTester,
  RootTester,
  joinClanTestData,
} from 'vote-aggregator-tests';
import {VoteAggregatorSdk} from '../../src';
import {PublicKey} from '@solana/web3.js';

describe('join_clan instruction', () => {
  it.each(joinClanTestData.filter(({error}) => !error))(
    'Works',
    async ({
      realm,
      root,
      member,
      memberVoterWeight,
      clan,
    }: JoinClanTestData) => {
      const realmTester = new RealmTester(realm);
      const rootTester = new RootTester({
        ...root,
        realm: realmTester,
      });
      const memberTester = new MemberTester({
        ...member,
        root: rootTester,
        membership: MemberTester.membershipTesters({
          membership: member.membership || [],
          root: rootTester,
        }),
      });
      const clanTester = new ClanTester({...clan, root: rootTester});
      const sdk = new VoteAggregatorSdk();
      expect(
        sdk.member.joinClanInstruction({
          rootData: rootTester.root,
          memberData: memberTester.member,
          clanAddress: clanTester.clanAddress,
          memberVwr: memberVoterWeight.address,
          payer: PublicKey.default,
        })
      ).resolves.toMatchSnapshot();
    }
  );
});
