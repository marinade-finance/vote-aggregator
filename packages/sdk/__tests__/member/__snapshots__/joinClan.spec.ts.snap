// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`join_clan instruction Works 1`] = `
TransactionInstruction {
  "data": {
    "data": [
      218,
      113,
      190,
      239,
      4,
      91,
      106,
      206,
    ],
    "type": "Buffer",
  },
  "keys": [
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": PublicKey(2T5rQ4Crtv4yxxs5myXQEPXyanwMnkimseRon7pzPRvf) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            39487464,
            54576989,
            66557893,
            32028752,
            49097065,
            55323534,
            16719996,
            37304883,
            35006655,
            352913,
            0,
          ],
        },
      },
    },
    {
      "isSigner": true,
      "isWritable": false,
      "pubkey": PublicKey(7RLeU6Hv1Wzd6GpYiVcm4FgH9DQpJxAeGpWhWgmBtrZu) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            17178032,
            6830781,
            46391999,
            2678306,
            10757367,
            7361388,
            32549969,
            23186142,
            39239674,
            1562769,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": PublicKey(DWCHP6gPZeE7mggkrFtbrcNjn7NhKSrwwhanrFJBecvZ) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            48915222,
            38003857,
            24532236,
            6656886,
            13446302,
            22462886,
            28567731,
            25551796,
            66743007,
            3043765,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": PublicKey(BzVuuiNRYp5NCkTBR7KpM64c3EJ5HjuHEzGqX1nbWWUN) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            5971983,
            27961988,
            44297099,
            40509199,
            18496327,
            12758528,
            20881507,
            36603530,
            44169095,
            2675690,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": PublicKey(4dX2rDrR3pkixrBry19YnU1jck6mtfGziESaNhWZHzrE) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            20094283,
            26401473,
            49006016,
            43956182,
            39951003,
            47399785,
            15468599,
            59865983,
            22239728,
            883579,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": PublicKey(BG6spgWJccBj7Y8VHfdhx4HXLgyzqecTfS5Rjx2wLhrT) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            59327884,
            64252112,
            27929104,
            21176826,
            61363990,
            62961203,
            11287407,
            33833427,
            29801411,
            2497742,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": PublicKey(FWFMJ1qeKUBqk9Mnuigf6Ln3PaPuBD69FLxZeB66YoCP) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            60591188,
            48147032,
            30662027,
            39731127,
            44725213,
            7359639,
            58198680,
            14712769,
            60233696,
            3530859,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": PublicKey(ApnCQDSeiRRbc4Bq9rNebfnyQdaNyM1suSbZ1vCmSUaF) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            21981212,
            24849885,
            28276057,
            14787534,
            18298572,
            56237326,
            59040209,
            36498233,
            26403464,
            2391461,
            0,
          ],
        },
      },
    },
  ],
  "programId": PublicKey(VoTaGDreyne7jk59uwbgRRbaAzxvNbyNipaJMrRXhjT) PublicKey {
    "_bn": BN {
      "length": 10,
      "negative": 0,
      "red": null,
      "words": [
        8873870,
        60943108,
        23935113,
        33099803,
        59555925,
        65803128,
        11262563,
        66819466,
        50754419,
        120882,
        0,
      ],
    },
  },
}
`;
